java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java

```diff
@@ -0,0 +1,126 @@
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S3510")
+public class HostnameVerifierImplementationCheck extends IssuableSubscriptionVisitor {
+
+  private static final String ISSUE_MESSAGE = "Do not unconditionally return true in this method.";
+
+  private static final TypeCriteria TYPE_CRITERIA_STRING = TypeCriteria.is("java.lang.String");
+  private static final TypeCriteria TYPE_CRITERIA_SSL_SESSION = TypeCriteria.is("javax.net.ssl.SSLSession");
+  private static final MethodMatcher VERIFY_METHOD_MATCHER = MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("javax.net.ssl.HostnameVerifier"))
+    .name("verify")
+    .addParameter(TYPE_CRITERIA_STRING)
+    .addParameter(TYPE_CRITERIA_SSL_SESSION);
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.kind().equals(Tree.Kind.METHOD)) {
+      checkMethodDefinition((MethodTree) tree);
+    } else if (tree.kind().equals(Tree.Kind.METHOD_INVOCATION)) {
+      ((MethodInvocationTree) tree).arguments().forEach(this::checkForLambdaImplementation);
+    } else {
+      ((NewClassTree) tree).arguments().forEach(this::checkForLambdaImplementation);
+    }
+  }
+
+  private void checkMethodDefinition(MethodTree tree) {
+    BlockTree blockTree = tree.block();
+    if (tree.symbol() != null && VERIFY_METHOD_MATCHER.matches(tree) && blockTree != null) {
+      checkBlock(blockTree);
+    }
+  }
+
+  private void checkForLambdaImplementation(ExpressionTree argument) {
+    if (argument.is(Tree.Kind.LAMBDA_EXPRESSION)) {
+      LambdaExpressionTree lambdaExpressionTree = ((LambdaExpressionTree) argument);
+      Tree lambdaBody = lambdaExpressionTree.body();
+      if (isHostnameVerifierSignature(lambdaExpressionTree)) {
+        if (lambdaBody.is(Tree.Kind.BLOCK)) {
+          checkBlock((BlockTree) lambdaBody);
+        } else if ((lambdaBody.is(Tree.Kind.PARENTHESIZED_EXPRESSION) || lambdaBody.is(Tree.Kind.BOOLEAN_LITERAL))
+          && isTrueLiteral((ExpressionTree) lambdaBody)) {
+          reportIssue(lambdaBody, ISSUE_MESSAGE);
+        }
+      }
+    }
+  }
+
+  private void checkBlock(BlockTree blockTree) {
+    BlockTree innerBlock = blockTree;
+    while (innerBlock.body().size() == 1 && innerBlock.body().get(0).is(Tree.Kind.BLOCK)) {
+      innerBlock = (BlockTree) innerBlock.body().get(0);
+    }
+
+    if (isReturnTrueStatement(innerBlock.body())) {
+      reportIssue(innerBlock.body().get(0), ISSUE_MESSAGE);
+    }
+  }
+
+  private static boolean isHostnameVerifierSignature(LambdaExpressionTree lambdaExpressionTree) {
+    List<VariableTree> parameters = lambdaExpressionTree.parameters();
+    return lambdaExpressionTree.symbolType().isSubtypeOf("javax.net.ssl.HostnameVerifier")
```

Comment:
if you have a lambda as a subtype of HostnameVerifier I don't think you need all the elements of the condition. (in other words, there is no way if this first part is true that any of the following might be false, and it is therefore uncoverable by unit test and useless) :  so all the following && can be safely removed I think. (please prove me wrong if I am ;) )